### ISSUE TITLE
Feature Request: Ping measurements also during load

### ISSUE THREAD

**Issue by zR-JB** [openspeedtest/Speed-Test#142]
Maybe never stop the ping measurement, and let it run continuously also during the upload and download phase. And then have a simple graph, or just the average during the different load scenarios. With this, one could investigate buffer bloat issues very easy!

Even ookla with speedtest.net had this feature for quite some time now:
![image](https://github.com/user-attachments/assets/22197711-4e19-4419-ae40-96c7bee267a2)

**Comment by openspeedtest** [comment]
I'll address this in a future version, but there are known limitations as outlined in https://github.com/openspeedtest/Speed-Test/issues/33. Otherwise, we'd need to overhaul the entire setup and potentially use something like webtransfer. I'll seriously consider this for the next major rewrite, which will support multiple protocols.

**Comment by zR-JB** [comment]
When looking at what is transferred during the speedtest.net test, I noticed that the downloads and uploads are also XHR, and the ping measurements are done via a websocket, so I think you are right.

![image](https://github.com/user-attachments/assets/c5cd4e25-10ce-49df-a024-a9d55dd7c39c)

**Comment by zR-JB** [comment]
So I did a bit more testing a made small POC with a simple node.js echo websocket server. And then performance.now() on the client side. Turns out Firefox does by default also only provide millisecond accuracy, even with performance.now(). Chromium based browsers, at least, have 100 μs. I think there is still a bit much overhead for extremely accurate results.

[websocket_latency_test.zip](https://github.com/user-attachments/files/16390566/websocket_latency_test.zip)

**Comment by zR-JB** [comment]
So here is what I found:

1. The Browser Web socket Implementation adds around 0.1 ms of latency compared to a native rust websocket client when running on the same machine
2. Using a Rust websocket-server instead of a node.js server only saves about 0.02 ms 
3. When being in the same network but another machine, with native rust <->rust websockets I get around 0.3 ms (between 0.24 and 0.4)
4. With this setup and the browser implementation client and rust websocket server, I get around 0.9ms. 
5. Real Latency via ICMP, in this scenario, is about 0.2 ms. 
6. So this is not ideal, but I am not sure if any other protocol can achieve anything better than that

**Comment by openspeedtest** [comment]
When you add browser extensions, web filters, antivirus software, and other busy tabs to the mix, you'll see a much higher RTT.

**Comment by zR-JB** [comment]
Yes, I think a websocket approach to measure latency similar to:

![grafik](https://github.com/user-attachments/assets/2c5e700b-645b-4e5b-b6df-3426e673e32e)
is probably still the best compromise

**Comment by ZMJGMADHPXWT** [comment]
Any news?

**Comment by zR-JB** [comment]
@ZMJGMADHPXWT I just built my own speed test suite to be honest

**Comment by netztier** [comment]
seasoned network operator here - not a developper; please bear with me using software develpment related terms in a totally noobish way, and please excuse the inability to provide (sample) code. 

I just set up my own hosted instance locally, liked it much, and also noticed the absence of "RTT under load".


**Thinking along these lines:**

Variety A: "inband" telemetry
- sacrifice one of the 6 possible concurrent connections for _inband telemetry_
- _inband telemetry_ to be optional, via Parameter in URL (paraphrasing ... `/?telemetry=inband`) 
- start the the 1st (or the 6th) thread/connection before the others, to do some unloaded RTT checking (like: ~5 sec), mimicking the way TWAMP does it, in a given interval (100, maybe 200ms). 
- if possible:  use a stratagy to build the telemetry connection in a way to a avoid a full TCP 3way-Handshake for every request. This might skew the results. 
- ramp up the 2nd- to 5th connections as needed to do the heavy lifting, while the 1st thread/connection keeps running and collecting data.

Variety B: "sideband" telemetry
- Parameter in ULR (... `/?telemetry=sideband`)
- use another endpont ("origin", I believe is the term for this?), such as port 3002 for an independent telemetry connection, to avoid running under the same browser context, not having to compete with 5 load-bearing connections.
- 3000 and 3001 are already in use, so adding a third port might not be that steep a requirement.
- use the same TWAMP logic as with inband, provided the browser's timestamp resolution ist good enough.

**TWAMP**:
- TWAMP originally operates on UDP, and also measures loss; loss can be glossed-over here (TCP would prevent/compensate it anyway), and I don't think UDP would actually be needed.
- TWAMP logic is roughly this: 
   - "sender" (Browser) embeds local timestamp T1 in whatever style of "request" it sends towards the reflector (and possibly a request ID)
   - "reflector" (here: openspeedtest server) determines T2 (own timestamp of request reception) and T3 (own timestamp of response processing)
   - "reflector" sends response containing T2 and T3
   - "sender" (Browser) receives packet, determines T4 (own timestamp of response reception)
   - Browser calculates (T4 -T1)  - (T3 - T2)  and has a "good enough" estimation of network RTT
   - time-sync NOT needed between "sender" and "reflector".
   - yes, there might be (undesirable) contributions of browser session handling and queing in the result - but see above: "good enough".
   - in very low-latency scenario, this might not be good enough to provide full accuracy, depending on timestamp resolution. 
   - buffer bloat effects are measured in many-tens-of-milliseconds, much less in sub-millisecond ranges, so "good enough" might just be, well, good enough.

**Output:** 
- Graph the RTT values over the test duration,
   - in their own graph beneath the bandwidth graph (same scale of X-Axis) 
   - or as a dotted-line overlay on the bandwitdh graph with it's own Y-Axis
   - or as something even better... (running out of ideas, here)

**Result:**
Users can instantly see when buffer bloat starts to hit along the network path to/from their server, even per direction, 
in a way none of the "big ones" (fast.com, speedtest.net) show it. 


Context: 
We do use built-in TWAMP on Fortigate Firewalls with SDWAN deployments, to continously assess the quality of the network paths. 

Edit:
Added ideas abot inband/sideband telemetry
